tinygrad/codegen/linearizer.py

```diff
@@ -518,26 +523,28 @@ def uop(self, uop:UOps, dtype:Optional[DType]=None, vin:Tuple[UOp, ...]=tuple(),
 
   def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]], do_reduce=False, loop_ctx=tuple()) -> List[UOp]:  # noqa: E501
     if x.op in BufferOps: return loaded_buffers[x.arg]
-    if x.op == UnaryOps.CAST: return [self.uop(UOps.CAST, x.arg[0], (u,), x.arg) if not isinstance(x.arg[0], ImageDType) else u for u in self.ast_parse(cast(LazyOp, x.src[0]), acc, offs, loaded_buffers)]  # noqa: E501
+    if x.op == UnaryOps.CAST: return [self.uop(UOps.CAST, dtype:=self.get_base_dtype(x.arg[0]), (u,), (dtype,x.arg[1])) for u in self.ast_parse(cast(LazyOp, x.src[0]), acc, offs, loaded_buffers)] # noqa: E501
     if x.op in ReduceOps and not do_reduce:
       assert offs is None, "not available if we aren't doing reduce"
       return acc
     # MULACC fusion. TODO: this is copied from Interpreted
-    if x.op == ReduceOps.SUM and x.src[0].__class__ is LazyOp and x.src[0].op == BinaryOps.MUL:
+    # TODO only doing MULACC in floats, ints can mulacc with autocast
+    # TODO should we also check self.reduceop being float? add tests for this
+    if x.op == ReduceOps.SUM and x.src[0].__class__ is LazyOp and x.src[0].op == BinaryOps.MUL and dtypes.is_float(self.get_base_dtype(get_lazyop_info(x).dtype)): # noqa: E501
```

Comment:
no new `# noqa: E501`. maybe use `isinstance` instead of `__class__ is`

Reply:
definitely, I'm planning to do an entire diff cleanup last thing once everything works